trunner/runners/wrappers.py

```diff
@@ -0,0 +1,339 @@
+            self.args,
+            cwd=self.cwd,
+            encoding='ascii'
+        )
+
+        self.read_output()
+        self.proc.wait()
+        if self.proc.exitstatus != 0:
+            logging.error('psu failed!\n')
+            raise Exception(' Loading plo using psu failed!')
+
+
+class Gdb(StandardWrapper):
+    """Wrapper for gdb-multiarch program"""
+
+    def __init__(self, target, file, script, cwd=None):
+        args = [f'{file}', '-x', f'{script}']
+        super().__init__(target, 'gdb-multiarch', args, cwd)
+
+    def read_output(self):
+        if is_github_actions():
+            logging.info('::group::Run gdb\n')
+
+        # When using docker, the additional newline may be needed
+        idx = self.proc.expect_exact(["(gdb) ", "Type <RET> for more"])
+
+        if idx == 1:
+            logging.info(self.proc.before)
+            self.proc.send('\r\n')
+            idx = self.proc.expect_exact("(gdb) ")
+
+        # plo will be loaded after passing the continue command in gdb
+        # TODO: investigate why send('c\r\n') does not work
+        self.proc.send('c')
+        self.proc.send('\r\n')
+
+        logging.info(f'{self.proc.before}\n')
+        self.proc.close()
+
+        if is_github_actions():
+            logging.info('::endgroup::\n')
+
+    def run(self):
+        with GdbServer() as gdbserv:
+            # Use pexpect.spawn to run a process as PTY, so it will flush on a new line
+            self.proc = pexpect.spawn(
+                self.progname,
+                self.args,
+                cwd=self.cwd,
+                encoding='ascii'
+            )
+
+            try:
+                self.read_output()
+            except (pexpect.TIMEOUT, pexpect.EOF):
+                exception = Color.colorify(f'\nGDB ERROR:', Color.BOLD)
+                exception += '\nGdb prompt not seen after executing loading script!'
+                exception = error_msg('gdb', exception, self.proc.before)
+                exception = error_msg('JlinkGdbServer', exception, gdbserv.output())
+
+                logging.info(exception)
+                sys.exit(1)
+            self.proc.close()
+
+    def close(self):
+        self.proc.close()
+        if self.proc.exitstatus != 0:
+            logging.error('gdb failed!\n')
+            raise Exception('Loading plo using gdb failed!')
+
+
+class PhoenixdError(Exception):
+    pass
+
+
+class Phoenixd(BackgroundWrapper):
+    """ Wrapper for phoenixd program"""
+
+    def __init__(
+        self,
+        target,
+        port,
+        dir='.',
+        cwd=None,
+        wait_dispatcher=True,
+        long_flashing=False
+    ):
+        if cwd is None:
+            cwd = boot_dir(target)
+        self.cwd = cwd
+        self.port = port
+        self.dir = dir
+        self.wait_dispatcher = wait_dispatcher
+        self.dispatcher_event = None
+        self.long_flashing = long_flashing
+        super().__init__()
+
+    def _reader(self):
```

Comment:
I've tested this approach, but realized that something like exit event seems to be necessary. It will be available in the next force push.

Reply:
EDIT: Using exit event finally wasn't working fine. I am not sure why, but when we only spawn `phoenixd` process using `pexpect.spawn` long copy command can't be completed successfully. It seems that `phoenixd` requires reading its output (?), because when I use reading line by line or simple expect(pexpect.EOF) everything works fine (in this case expect is much faster).

I could expect EOF, like `ret=0` before, but there was no way to catch output if EOF didn't occur - e.g . phoenixd got stuck (we could set timeout, but we even don't know what generic value would be good - for long copy it should be ~140s, for shorter copy (imxrt) ~60s and for app command a few seconds). So it would neither be a good solution.

Finally I've decided to use logfile utility. After `pexpect.spawn` logfile is opened and in _reader thread I am expecting `EOF`, so the`phoenixd` output is constantly read.  The whole phoenixd output I can read from the logfile in output() method. If everything had gone properly _reader thread catches EOF and ends, if not it would be killed from _exit(), then logfile is closed and eventually opened and read if output was needed (output() method).

The changes will be available soon, I hope it's understandable